Add Preloading responsive images blog post (#1555)
GoogleChrome/web.dev: src/site/content/en/blog/preload-responsive-images/index.md

---
layout: post
title: Preloading responsive images

> subhead: Learn about new and exciting possibilities for preloading responsive images to ensure great user experience.

Reviewer: ```suggestion
subhead: Starting in Chrome 73, link rel="preload" and responsive images can be combined in order to load images faster.
```

authors:
  - yoavweiss
description: |
  Learn about new and exciting possibilities for preloading responsive images to ensure great user experience.
hero: hero.jpg

> alt: A wall with a bunch of images frames in different sizes.

Reviewer: ```suggestion
alt: A wall with a bunch of image frames in different sizes.
```

date: 2019-09-27
tags:
  - post # post is a required tag for the article to show up in the blog.
  - fast
  - images
---

This article gives me an opportunity to discuss two of my favorite things: responsive images *and* preload. As someone who was heavily involved in developing both of those features, I'm super excited to see them working together!

## Responsive images overview

> Suppose you're browsing the web on a screen that's 300 pixels wide, and the page just requested an image that's 1500 pixels wide. That page just wasted a lot of your cellular data because your screen can't do anything with all of that extra resolution. Ideally, the browser should fetch a version of the image that's just a *little* wider than your screen size, say 325 pixels. This ensures a high-resolution image without wasting data. And, even better, the image will load faster. [Responsive images](/serve-responsive-images/#serve-multiple-image-versions) enable browsers to fetch different image resources to different devices. You need to save multiple dimensions for each image and specify them in the `srcset` attribute. The `w` value tells the browser the width of each version. Depending on the device, the browser can choose the appropriate one:

Reviewer: ```suggestion
Suppose you're browsing the web on a screen that's 300 pixels wide, and the page just requested an image that's 1500 pixels wide. That page just wasted a lot of your cellular data because your screen can't do anything with all of that extra resolution. Ideally, the browser should fetch a version of the image that's just a *little* wider than your screen size, say 325 pixels. This ensures a high-resolution image without wasting data. And, even better, the image will load faster. [Responsive images](/serve-responsive-images/#serve-multiple-image-versions) enable browsers to fetch different image resources to different devices. If you don't use an [image CDN](/image-cdns/) need to save multiple dimensions for each image and specify them in the `srcset` attribute. The `w` value tells the browser the width of each version. Depending on the device, the browser can choose the appropriate one:
```

Adds a mention of image CDNs for solving the problem of creating multiple sizes of each image

```html
<img src="small.jpg"

     srcset="small.jpg 500w,

            medium.jpg 1000w,

            large.jpg 1500w"

            alt="…">

> ```

Reviewer: Please remove the extra spacing in this code

![image](https://user-images.githubusercontent.com/4713486/65640224-b2da7980-df9e-11e9-9695-bad8e932303e.png)

Responsive images and preload have been available for the last few years, but at the same time something was missing: there was no way to preload responsive images. [Starting in Chrome 73](https://developers.google.com/web/updates/2019/03/nic73#more), the browser can preload the right variant of responsive images specified in `srcset` before it discovers the `img` tag!

Depending on your site's structure, that could mean significantly faster image display! We ran tests on a site that uses Javascript to lazy-load responsive images. Preloading resulted in images loading 1.2 seconds faster.

{% Aside %}

Responsive images are [supported in all modern browsers](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#Browser_compatibility) while preloading them is [supported only in Chromium-based browsers](https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content#Browser_compatibility).

{% endAside %}

## `imagesrcset` and `imagesizes`

To preload responsive images, new attributes were recently added to the `<link>` element: `imagesrcset` and `imagesizes`.  They are used with `<link rel="preload">` and match the `srcset` and `sizes` syntax used in `<img>` element.

> If you want to preload a responsive image specified with:

Reviewer: ```suggestion
For example, if you want to preload a responsive image specified with:
```

Of course, nothing captures the visual difference quite like a filmstrip comparison:
<figure class="w-figure">
<img src="./example-3.png" alt="Screenshot of WebPageTest filmstrip comparison showing preloaded images are displayed about 1.5 seconds faster.">
<figcaption class="w-figcaption">The filmstrip shows that images arrive significantly faster when preloaded, resulting in a hugely-improved user experience.</figcaption>
<figure> 

## Preload and `<picture>`?

If you're familiar with responsive images, you may be wondering "What about [`<picture>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture)?".

> We're talking about adding a preload equivalent for `srcset` and `sizes`, but not the `<picture>` element, which tackles the ["art direction"](/codelab-art-direction/) use-case.

Reviewer: ```suggestion
The Web Performance Working Group is talking about adding a preload equivalent for `srcset` and `sizes`, but not the `<picture>` element, which tackles the ["art direction"](/codelab-art-direction/) use-case.
```

Why is this use-case being "neglected"?

While there's interest in solving that use case as well, there are still a number of [technical issues to sort out](https://calendar.perfplanet.com/2018/how-the-sausage-is-made-webperfwg-meeting-summary/) which means that a solution here would have significant complexity. On top of that, it seems like for the most part, the use-case can be addressed today, even if in a hacky way (see below). 

Given that, the Web Performance WG decided to ship `srcset` first and see if the demand for equivalent `picture` support arises.

If you do find yourself in a position to preload `<picture>` you may be able to use the following technique as a workaround.

Given the following scenario:

```html
<picture>
    <source src="small_cat.jpg" media="(max-width: 400px)">

    <source src="medium_cat.jpg" media="(max-width: 800px)">

    <img src="huge_cat.jpg">
</picture>

> ```

Reviewer: Please remove the extra spacing

![image](https://user-images.githubusercontent.com/4713486/65641116-cab2fd00-dfa0-11e9-8535-e316d9852948.png)

The `<picture>` element's logic (or the image source selection logic, to be precise), would be to go over the `media` attributes of the `<source>` elements in order, find the first one that matches, and use the attached resource.

Because responsive preload has no notion of "order" or "first match", the breakpoints need to be translated into something like:

```html
<link rel="preload" href="small_cat.jpg" as="image" media="(max-width: 400px)">

<link rel="preload" href="medium_cat.jpg" as="image" media="(min-width: 400.1px) and (max-width: 800px)">

<link rel="preload" href="large_cat.jpg" as="image" media="(min-width: 800.1px)">

> ```

Reviewer: This extra spacing is more debatable but let's just remove it to be consistent
